### DISCUSSION TITLE
Single signon

### DISCUSSION THREAD

**Discussion by jcalfee** [requarks/wiki#5149] [Help / Questions] [answered]
What is the best strategy for single sign on?  We have an authentication server where they will already be logged in.  Just need to get them into the self-hosted wiki where they will automatically be setup and signed in.  A default group is essential, but ideally I would like the authentication server to provide a well known group at the wiki for each user.

LDAP may do it and has a default group, but it does not look like the auto-login will work.  JWT does not seem to setup accounts or support groups.

**Comment by jcalfee** [comment]
I didn't think about combining them.  Looks like I could add them in a LDAP server then use oAuth or JWT to log them in.  Could this passport.js thing make it easier?

**Comment by jcalfee** [comment]
Same question: https://github.com/requarks/wiki/discussions/4665

**Comment by jcalfee** [comment]
This is as seamless as it gets on my local machine.  One server may be used to authenticate then hand off to the Wiki where the new account will be created or the existing account gets authenticated.  If you logout from an account created using oAuth2, you'll go back to the oAuth2 server; if you logout from an account created in the Wiki's local database, you'll go back to the wiki's login screen.

**I have not yet figured out how to trigger a refresh of the OAuth2 token, so there is no way for the oauth2orize server to remove access from the Wiki if they remain logged in.  The JWT timeout value is different, it does not interact with oAuth2.**  If you find a solution for that, please let me know.

Clone https://github.com/jaredhanson/oauth2orize .. In `routes/sites.js` change `successReturnToOrRedirect: '/'` to the callback URL from the Wiki's Admin Authentication "Generic OAuth2" screen (example: `successReturnToOrRedirect: 'http://localhost:3000/login/173b3c95-5dad-4d8c-b217-8fbb8281c636/callback'`, your UUID will differ).  Also in the same file, change `module.exports.logout ... response.redirect('/')` to `response.redirect('/login')`.  In `routes/user.js` in `module.exports.info` include `email: request.user.email` and in `db/users.js` add `email` to each user in `const users`.  Adding console logs to the `db/*.js` files is very useful.  Run this server with the a free port `PORT=3012 node app.js`.  In the Wiki, Admin, General, make sure your site URL is correct, that will be used.

Find the Client ID of the trusted client `xyz123` in `db/clients.js` and use that in the Wiki's Admin Authentication "Generic OAuth2" screen.  Make sure the Wiki's oAuth2 configuration has self registration enabled, without that the new account on the Wiki will not be created.

![image](https://user-images.githubusercontent.com/204121/163452353-049fe419-7841-4cbc-b510-e0281448e8d5.png)

Your uuid in the callback will be different.
![image](https://user-images.githubusercontent.com/204121/163452585-12cf5b67-5643-44ef-917d-a50bf1890887.png)

Because I used a client id from `db/clients.js` marked as trusted, the `/dialog/authorize` path is not applicable.

**Comment by jcalfee** [comment]
For refresh support, it looks like a patch is needed in `server/modules/authentication/oauth2/authentication.js` (see https://docs.requarks.io/en/dev/authentication) and include  https://github.com/fiznool/passport-oauth2-refresh .. Also an expiration value has to be added in the auth server and checked later in the wiki patch triggering the actual refresh call.

**Reply by jcalfee** [reply]
Recheck patch: https://github.com/requarks/wiki/discussions/4665#discussioncomment-2600588

**Comment by NGPixel** [comment, marked as answer]
There's no automated SSO support at the moment in Wiki.js.

**Comment by fourgates** [comment]
I was able to get this working by using cookies. Our auth server and wikijs are on different subdomains and can share a cookie named "jwt". We had to update the jwt token we were using for our app to include the info that wikijs was looking for.  

e.g.

app.domain.com and wiki.domain.com

**Reply by jcalfee** [reply]
I assume you can renew the cookie when it expires and the wikijs can do this too.  To what extent though?  From the external domain, can you turn off accounts, update email or profile, create accounts that don't exist?  Or did you just need to get them in and not need to do the rest?  I don't need an example, but someone might like to see something that can become a library esp if it can handle a more robust set of use-cases.

**Reply by fourgates** [reply]
1. yes our app renews the cookie when it is set to expire in a configured amount of time. its the same token we use for our main app so it follows the same policies. 
2. we do all the user management (at your own risk!) inside our external domain. under the hood its just a SQL database. wikijs handles the use case of throwing an exception for a cookie/token of an inactive user. You can still do this inside wikijs as well. 
3. our goal was to have our external domain automate creating users and groups in the wiki once they are CRUD in the main app. we then iframe the wiki into our app on a sub-domain.

the only issue we have had so far is that occasionally the wiki returns its own cookie (domain=wiki.mydomain.com) which conflicts with the external app cookie (domain=.mydomain.com). users end up with two cookies (one with each domain). 

Clearing cookies or manually removing the wiki cookie (wiki.mydomain.com) solves the issues. This seldomly happens so we haven't spent too much time on it. I suspect it has something to do with the cookie/token renew configuration between the two apps. 

**NOTE: This functionality is not supported. do so at your own risk**
